flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeGeneralization.java

```diff
@@ -0,0 +1,631 @@
+					resultType = createCommonApproximateNumericType(resultType, type);
+				} else if (hasFamily(resultType, EXACT_NUMERIC)) {
+					// the result was exact so far
+					if (typeRoot == DECIMAL) {
+						// in case of DECIMAL we enforce DOUBLE
+						resultType = new DoubleType();
+					} else {
+						// enforce an approximate result
+						resultType = type;
+					}
+				} else {
+					return null;
+				}
+			}
+			// for TIME
+			else if (hasFamily(type, TIME)) {
+				if (hasFamily(resultType, TIME)) {
+					resultType = new TimeType(combinePrecision(resultType, type));
+				} else {
+					return null;
+				}
+			}
+			// for TIMESTAMP
+			else if (hasFamily(type, TIMESTAMP)) {
+				if (hasFamily(resultType, TIMESTAMP)) {
+					resultType = createCommonTimestampType(resultType, type);
+				} else {
+					return null;
+				}
+			}
+			// for day-time intervals
+			else if (typeRoot == INTERVAL_DAY_TIME) {
+				resultType = createCommonDayTimeIntervalType(
+					(DayTimeIntervalType) resultType,
+					(DayTimeIntervalType) type);
+			}
+			// for year-month intervals
+			else if (typeRoot == INTERVAL_YEAR_MONTH) {
+				resultType = createCommonYearMonthIntervalType(
+					(YearMonthIntervalType) resultType,
+					(YearMonthIntervalType) type);
+			}
+			// other types are handled by findCommonCastableType
+			else {
+				return null;
+			}
+		}
+
+		// NULL type only
+		if (resultType == null && hasNullType) {
+			return new NullType();
+		}
+
+		return resultType;
+	}
+
+	private static boolean areSimilarTypes(LogicalType left, LogicalType right) {
+		// two types are similar iff they can be the operands of an SQL equality predicate
+
+		// similarity based on families
+		if (hasFamily(left, CHARACTER_STRING) && hasFamily(right, CHARACTER_STRING)) {
+			return true;
+		} else if (hasFamily(left, BINARY_STRING) && hasFamily(right, BINARY_STRING)) {
+			return true;
+		} else if (hasFamily(left, NUMERIC) && hasFamily(right, NUMERIC)) {
+			return true;
+		} else if (hasFamily(left, TIME) && hasFamily(right, TIME)) {
+			return true;
+		} else if (hasFamily(left, TIMESTAMP) && hasFamily(right, TIMESTAMP)) {
+			return true;
+		}
+		// similarity based on root
+		return left.getTypeRoot() == right.getTypeRoot();
+	}
+
+	private static @Nullable LogicalType findExactlySameType(List<LogicalType> normalizedTypes) {
+		final LogicalType firstType = normalizedTypes.get(0);
+		for (LogicalType type : normalizedTypes) {
+			if (!type.equals(firstType)) {
+				return null;
+			}
+		}
+		return firstType;
+	}
+
+	private static @Nullable LogicalType findCommonTypePattern(LogicalType resultType, LogicalType type) {
+		if (hasFamily(resultType, DATETIME) && hasFamily(type, INTERVAL)) {
+			return resultType;
+		} else if (hasFamily(resultType, INTERVAL) && hasFamily(type, DATETIME)) {
+			return type;
+		} else if ((hasFamily(resultType, TIMESTAMP) || hasRoot(resultType, DATE)) && hasFamily(type, EXACT_NUMERIC)) {
```

Comment:
Out of curiosity why doesn't it work for `TIME_WITHOUT_TIME_ZONE`?

If I understand correctly: `DATETIME - TIME_WITHOUT_TIME_ZONE = TIMESTAMP + DATE`

Reply:
I added the following comment:
```
// for "DATETIME + EXACT_NUMERIC", EXACT_NUMERIC is always treated as an interval of days
// therefore, TIME + EXACT_NUMERIC is not supported
```

This is also in sync with Oracle's behavior.